pulsar-broker/src/test/java/org/apache/pulsar/client/api/KeySharedSubscriptionTest.java

```diff
@@ -1630,4 +1629,66 @@ public void testContinueDispatchMessagesWhenMessageDelayed() throws Exception {
         log.info("Got {} other messages...", sum);
         Assert.assertEquals(sum, delayedMessages + messages);
     }
+
+    /**
+     * This test validates key-shared subscription should not get stuck and if one of the consumer dies then broker
+     * should redeliver those messages to the new consumer instead stop dispatching messages to all consumers.
+     * 
+     * @throws Exception
+     */
+    @Test
+    public void testKeySharedMessageRedeliveryWithoutStuck()
```

Comment:
> can we remove the blocker if there is no concern about the PR and can we merge it as it's really needed for various production systems.

I created a channel to trace the context more easily.

---

### Menu:
- 1.The concerns of the test `testKeySharedMessageRedeliveryWithoutStuck`
- 2.Clear the problem the PR is supposed to solve
- 3.This PR will bring issues

### 1.The concerns of the test `testKeySharedMessageRedeliveryWithoutStuck`

The test works like this:
- create `consumer1`
- send `10` msgs
- create `consumer2`
- send `10` msgs
- create `consumer3`
- redeliver all messages of `consumer2`
- send `10` msgs
- close `consumer1`
- receive all messages for `consumer2`
- receive all messages for `consumer3`

#### Concens
- The Step "redeliver all messages of `consumer2`" is not meaningful. It did nothing.
- We should run "receive all messages for `consumer2`" and "receive all messages for `consumer3`" in different threads, so they could receive all messages(Acknowledge received messages in time).

#### What did the test(without my suggestion) prove
- It proved that one stuck consumer can stuck another one.
- It does not prove that ultimately there will be messages that will not be consumed(see `Concern 2` above).

### 2.Clear the problem the PR is supposed to solve
- You just want to improve performance and prevent idle reading, right? (The test can not prove the subscription would be stuck, see `Concern 2` above ).

### 3.This PR will bring issues

| No. | Consumer 1 | Consumer 2 | Consumer 3 | Consumer 4 |
| --- | --- | --- | --- | --- |
| stat | handling `k1,k2`, `recent-join: null`  | handling `k3,k4`, `recent-join: null` |
| 1 | received `M1(k1), M2(k2)`| |
| 2 | | | added | 
| description | | | assigned `k2` which from Consumer 1 |
| stat | handling `k1` | handling `k3,k4`| handling `k2`, `recent-join: M2` |
| 3 | | received `1000` messages (`M3(k3)...M1002(k3)`) |
| 4 | | closed | 
| description | assigned `k3` which from Consumer 2 | |  assigned `k4` which from Consumer 2 |
| stat | handling `k1,k3`, `recent-join: null` | | handling `k2, k4`, `recent-join: M2` |
| 5 | received `M3(k3)...M1000(k3)`, the incoming queue is full now.|
| 6 | | | | added |
| description |  | |  | assigned `k3` which from Consumer 1 |
| state |handling `k1`, `recent-join: null` | | handling `k2, k4`, `recent-join: M2` | handling `k3`, `recent-join: M1002` |
| 7 | | | | Since the message `M2` has not been acked, `M1001(k3)...M1002(k3)` will be filter out|

In Step 7, the code you're removing alleviates some of the message-ordering problems, even if these codes don't solve all of the cases. https://github.com/apache/pulsar/pull/20776 is trying to solve all the cases of this problem. But if you remove this mechanism before https://github.com/apache/pulsar/pull/20776 is complete, then the out-of-order problem will be worse.